### PR TITLE
Enable/disable bootstrapping and customize store location

### PR DESCRIPTION
**Pull request by alalazo**
This PR adds a new command:
```console
$ spack bootstrap --help
usage: spack bootstrap [-h] SUBCOMMAND ...

manage bootstrap configuration

positional arguments:
  SUBCOMMAND
    enable    enable bootstrapping
    disable   disable bootstrapping
    reset     reset bootstrapping configuration to Spack defaults
    root      get/set the root bootstrap directory

optional arguments:
  -h, --help  show this help message and exit
```
to manage bootstrapping configuration. Currently the command can enable/disable bootstrapping:
```console
$ spack bootstrap enable
$ spack bootstrap disable
```
get/set a new root folder for bootstrapping:
```console
$ spack bootstrap root
/Users/culpo/.spack/bootstrap
$ spack bootstrap root "\$spack/opt/bootstrap"
/Users/culpo/PycharmProjects/spack/opt/bootstrap
``` 
and reset configuration to Spack's default:
```console
$ spack bootstrap reset
==> Bootstrapping configuration is being reset to Spack's defaults. Current configuration will be lost.
Do you want to continue? [Y/n]
``` 
Each setting is stored in a new `bootstrap.yaml` file that contain all the configuration settings related to bootstrapping.

### CODE REVIEW

**Comment by alalazo** [comment]
@jjellio I think you may be interested in this PR. It permits to set:
```yaml
config:
  bootstrap:
    enable: false
```
to prevent the bootstrapping procedure kicking-in accidentally.

**Comment by alalazo** [comment]
@tgamblin @becker33

**Comment by sethrj** [comment]
Yes please @alalazo ! I'm not sure I understand the need for a separate top-level file for bootstrapping alone (seems to me this would be a good fit inside "config") but I'm a fan. Frankly I'd like an option to prevent `~/.spack` from ever existing, but that's outside the scope of this PR 😄 

Thanks!

**Comment by alalazo** [comment]
> (seems to me this would be a good fit inside "config")

It was implemented in `config.yaml` initially, but we figured out that to support bootstrapping from binaries we'll need more configuration entries and we may add even more in the future, so decided to keep everything related to bootstrapping into its own YAML file.

File: lib/spack/spack/bootstrap.py [hunk 1]
```diff
@@ -226,6 +228,23 @@ def ensure_bootstrap_configuration():
...
+    """Path to the store used for bootstrapped software"""
+    enabled = spack.config.get('bootstrap:enable', True)
+    if not enabled:
+        msg = ('bootstrapping is currently disabled. '
+               'Modify config.yaml to enable it')
+        raise RuntimeError(msg)
+
+    bootstrap_root_path = spack.config.get(
+        'bootstrap:root', None
+    ) or spack.paths.user_bootstrap_path
```
Comment (becker33) [inline lib/spack/spack/bootstrap.py]:
```suggestion
    bootstrap_root_path = spack.config.get(
        'bootstrap:root', spack.paths.user_bootstrap_path
    )
```

File: lib/spack/spack/bootstrap.py [hunk 2]
```diff
@@ -226,6 +228,23 @@ def ensure_bootstrap_configuration():
                         yield
 
 
+def store_path():
+    """Path to the store used for bootstrapped software"""
+    enabled = spack.config.get('bootstrap:enable', True)
+    if not enabled:
+        msg = ('bootstrapping is currently disabled. '
+               'Modify config.yaml to enable it')
```
Comment (becker33) [inline lib/spack/spack/bootstrap.py]: This message is outdated.

File: lib/spack/spack/cmd/bootstrap.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    # Set to True if we called "enable", otherwise set to false
+    value = str(args.subcommand == 'enable').lower()
+    scope, config_args = args.scope, []
+    if scope:
+        config_args.append('--scope={0}'.format(scope))
+    config_args.extend([
+        'add', 'bootstrap:enable:{0}'.format(value)
+    ])
+    _config(*config_args)
```
Comment (becker33) [inline lib/spack/spack/cmd/bootstrap.py]: Why use the config command interface here instead of the internal interface?

```suggestion
    # Set to True if we called "enable", otherwise set to false
    value = str(args.subcommand == 'enable').lower()
    spack.config.set('bootstrap:enable', value, scope=args.scope)
```

File: lib/spack/spack/cmd/bootstrap.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        )
+        if not ok_to_continue:
+            raise RuntimeError('Aborting')
+
+    for scope in spack.config.config.file_scopes:
+        if 'defaults' in scope.path:
+            continue
+        bootstrap_yaml = os.path.join(scope.path, 'bootstrap.yaml')
+        if os.path.exists(bootstrap_yaml):
+            os.remove(bootstrap_yaml)
```
Comment (becker33) [inline lib/spack/spack/cmd/bootstrap.py]: This won't handle the case where the bootstrap configuration is coming from an environment. What about 
```suggestion
    for scope in spack.config.config.file_scopes:
        if 'defaults' in scope.path:
            continue
        spack.config.set('bootstrap', {}, scope=scope)
```

File: lib/spack/spack/cmd/bootstrap.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def _root(args):
+    if args.path:
+        spack.config.set('bootstrap:root', args.path, scope=args.scope)
+
+    root = spack.config.get('bootstrap:root', default=None, scope=args.scope)
+    if root:
+        root = spack.util.path.canonicalize_path(root)
+    print(root)
```
Comment (becker33) [inline lib/spack/spack/cmd/bootstrap.py:100]: Should this still print even if we're setting the path?

**Review by becker33 — CHANGES_REQUESTED** [review]
Didn't review the tests yet.

File: lib/spack/spack/cmd/bootstrap.py (hunk 5 above)
Comment (alalazo) [reply lib/spack/spack/cmd/bootstrap.py:100]: Yeah, I think it should still print as a confirmation of the value that was set. The result is like shown in the description at the top.

File: lib/spack/spack/cmd/bootstrap.py (hunk 4 above)
Comment (alalazo) [reply lib/spack/spack/cmd/bootstrap.py]: Is there a clean way to use:
```python
spack.config.set('bootstrap', {}, scope=scope)
```
only if the config is from an environment file? What I don't like of this solution (which is good functionally, i.e. spack would behave as if things are removed) is that it will create files with empty configurations in each scope. I also wonder if this will fail in case we don't have write permissions for a scope that otherwise would not be touched.

Comment (becker33) [reply lib/spack/spack/cmd/bootstrap.py]: Well, we could first check `if spack.config.get('bootstrap', scope=scope)` to check whether any change is necessary for each scope. And all env scopes are guaranteed to start with `env` IIRC, so we could check that and remove files in the non-env case.

File: lib/spack/spack/bootstrap.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/bootstrap.py]: Fixed in 69a2487

File: lib/spack/spack/bootstrap.py (hunk 2 above)
Comment (alalazo) [reply lib/spack/spack/bootstrap.py]: Fixed in 69a2487

File: lib/spack/spack/cmd/bootstrap.py (hunk 3 above)
Comment (alalazo) [reply lib/spack/spack/cmd/bootstrap.py]: Fixed in 69a2487

File: lib/spack/spack/bootstrap.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/bootstrap.py]: The "or" idiom is necessary to be permissive and allow explicit "null" values for "root". See 206904d Let me know if this needs further thinking.

File: lib/spack/spack/cmd/bootstrap.py (hunk 4 above)
Comment (alalazo) [reply lib/spack/spack/cmd/bootstrap.py]: See 555cd16

**Comment by alalazo** [comment]
@becker33 Ready for a second review

File: lib/spack/spack/bootstrap.py (hunk 1 above)
Comment (becker33) [reply lib/spack/spack/bootstrap.py]: Shouldn't the validation require a string, in which case the value can't be explicitly `None` anyway?

Comment (alalazo) [reply lib/spack/spack/bootstrap.py]: Maybe it's overly defensive, since we ship a sensible default, but the rationale here was to protect against either the empty string or the config file being deleted. I changed the implementation to be a bit less defensive but simpler in 783f9b74e

File: lib/spack/spack/test/cmd/bootstrap.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+@pytest.mark.parametrize('scope', [
+    None, 'site', 'system', 'user'
+])
+def test_root_get_and_set(mutable_config, scope):
+    scope_args, path = [], '/scratch/spack/bootstrap'
+    if scope:
+        scope_args = ['--scope={0}'.format(scope)]
+
+    _bootstrap('root', path, *scope_args)
+    out = _bootstrap('root', *scope_args, output=str)
```
Comment (becker33) [inline lib/spack/spack/test/cmd/bootstrap.py:40]: Don't the scope args have to come before the path? If this is passing, it makes me worried about the validity of this test.

Comment (alalazo) [reply lib/spack/spack/test/cmd/bootstrap.py:40]: I don't think the position matters, since the positional argument is optional:
```console
$ spack bootstrap root "\$spack/opt/bootstrap" --scope=user
/home/culpo/PycharmProjects/spack/opt/bootstrap

$ cat ~/.spack/bootstrap.yaml 
bootstrap:
  root: $spack/opt/bootstrap
```
See also [this answer](https://stackoverflow.com/a/4480202/771663) where they parse `somedir -v`.

File: lib/spack/spack/cmd/bootstrap.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        ]
+        ok_to_continue = llnl.util.tty.get_yes_or_no(
+            ''.join(msg), default=True
+        )
+        if not ok_to_continue:
+            raise RuntimeError('Aborting')
+
+    for scope in spack.config.config.file_scopes:
+        # The default scope should stay untouched
+        if 'defaults' in scope.path:
```
Comment (becker33) [inline lib/spack/spack/cmd/bootstrap.py:74]: This should check for `scope.name`, since someone could put all of spack in a path that includes `default`.

File: lib/spack/spack/cmd/bootstrap.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            continue
+
+        # If we are in an env scope we can't delete a file, but the best we
+        # can do is nullify the corresponding configuration
+        if 'env' in scope.path and spack.config.get('bootstrap', scope=scope.name):
```
Comment (becker33) [inline lib/spack/spack/cmd/bootstrap.py]: `scope.name.startswith('env:')`, the scope path might not have `env` in it at all.

Comment (becker33) [inline lib/spack/spack/cmd/bootstrap.py:79]: `if scope.name.startswith('env:')`, the path might not include 'env'

File: lib/spack/spack/cmd/bootstrap.py (hunk 7 above)
Comment (becker33) [inline lib/spack/spack/cmd/bootstrap.py]: The user could have other scopes that have "defaults" in them, especially if they put spack in a path that has "defaults" in it.
```suggestion
        if scope.name == 'defaults':
```

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/cmd/bootstrap.py (hunk 6 above)
Comment (becker33) [reply lib/spack/spack/test/cmd/bootstrap.py:40]: Good to know it's only required positional arguments, not optional ones, that have to have the options first.

File: lib/spack/spack/cmd/bootstrap.py (hunk 8 above)
Comment (alalazo) [reply lib/spack/spack/cmd/bootstrap.py]: Done in d0ed0f2

File: lib/spack/spack/cmd/bootstrap.py (hunk 7 above)
Comment (alalazo) [reply lib/spack/spack/cmd/bootstrap.py]: Done in d0ed0f2

**Comment by alalazo** [comment]
@becker33 I think all the comments have been addressed, let me know if you see anything else that need changes.

**Comment by alalazo** [comment]
@becker33

File: lib/spack/spack/cmd/bootstrap.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        # can do is nullify the corresponding configuration
+        if (scope.name.startswith('env') and
+                spack.config.get('bootstrap', scope=scope.name)):
+            spack.config.set('bootstrap', {}, scope=scope.name)
+            continue
+
+        # If we are outside of an env scope delete the bootstrap.yaml file
+        bootstrap_yaml = os.path.join(scope.path, 'bootstrap.yaml')
+        if os.path.exists(bootstrap_yaml):
+            os.remove(bootstrap_yaml)
```
Comment (becker33) [inline lib/spack/spack/cmd/bootstrap.py]: Should we move the file to `bootstrap_yaml + '.bk'` instead of deleting it? Then whichever one was there last is available.

File: lib/spack/spack/schema/bootstrap.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Schema for bootstrap.yaml configuration file."""
+
+properties = {
+    'bootstrap': {
+        'type': 'object',
+        'properties': {
+            'enable': {'type': 'boolean'},
```
Comment (becker33) [inline lib/spack/spack/schema/bootstrap.py:11]: I think we should make this a list of enum values so that we can extend it later

```
bootstrap:
  enable: []
```
```
  enable: [source]
```
```
  enable: [binary]
```
```
  enable: [binary, source]
```

**Review by becker33 — CHANGES_REQUESTED** [review]
Found a couple more small things, should be pretty simple changes.

File: lib/spack/spack/cmd/bootstrap.py (hunk 9 above)
Comment (alalazo) [reply lib/spack/spack/cmd/bootstrap.py]: Done in e4dcb53

File: lib/spack/spack/schema/bootstrap.py (hunk 10 above)
Comment (alalazo) [reply lib/spack/spack/schema/bootstrap.py:11]: I considered this design at the beginning and I opted for a Boolean variant. One reason for that is to keep things simple and have a workflow for bootstrapping that always privileges binaries (if there's a trusted source) over sources i.e. I wanted to avoid having to support:
```yaml
bootstrap:
  enable: ['source', 'binary']
```
Another reason is that `binary` isn't a single "entity", but needs to be expanded into a list itself (e.g. Github Actions, LLNL build farm, custom mirrors, ... ). So I was thinking to keep a single list somewhere in this configuration with more structured entries, like:
```yaml
bootstrap:
  binary-providers:
  - name: github-actions
    description: 'Binary mirrors built on the cloud'
    type: buildcache
    trusted: unknown
  - name: llnl
    description: 'Binary mirrors from the LLNL build farm'
    type: buildcache
    trusted: unknown
```
So far I didn't consider disabling bootstrapping from sources as a last resort, but we can add that if that's a valid use case.

This is to say that I'm obviously open to discussing this, but would it be fine with you to merge as is for the time being and have the discussion when we'll move to #22720 and add also the trust mechanism for binaries? I think in that PR we'll be in a better position to try different solutions if need be.

Comment (becker33) [reply lib/spack/spack/schema/bootstrap.py:11]: I'm worried that if we merge this, we'll be stuck with backwards compatibility code (or an `update` method for `bootstrap.yaml`) when we change it later. I hadn't actually considered the issue of `enable: [source, binary]` though, I think I was considering the entries to be a set and Spack would always prefer binaries if it's an option, and fall back to source if there was no binary trusted option for that particular dependency. So I was thinking something like

```
bootstrap:
  enable: [source, binary]
  binary_providers:
  - name: mymirror
    description: ...
    type: # I'm not sure what this does
    trusted: true
  - name: spack-llnl
    ...
    trusted: false
  - name: spack-github
    ...
    trusted: unknown
```
Would mean for each dependency.
1. prompt whether we should trust `spack-github` (assume they say yes for this example)
2. Check for it in mymirror, if it's not there or install fails then
3. Check for it in spack-github, if it's not there or install fails then
4. Build from source

Comment (alalazo) [reply lib/spack/spack/schema/bootstrap.py:11]: Yeah, what I would do there is something more like:
```yaml
bootstrap:
  enable: true  # This just means I want to try bootstrapping
  providers: # Here I have the methods in order with enough info to use them
  - name: mymirror
    description: ...
    type: buildcache  # This one bootstraps from a Spack buildcache
    trusted: true
```
If we want to make bootstrapping from source explicit we can add an entry like:
```yaml
- name: spack-build
  description: "From source build of the missing dependency"
  type: source_build
  trusted: true
```
The entry `type` is meant to identify the kind of source we are using. At the moment we have only `buildcache` and `source_build` (names are just placeholders, of course) but we may figure out other ways later e.g. install something on the fly with `pip` in a virtual env.

Basically the idea is to use enable to check whether we want to do bootstrap or not. Then use a list of objects to express the order in which we want to try things.

Comment (alalazo) [reply lib/spack/spack/schema/bootstrap.py:11]: > I'm worried that if we merge this, we'll be stuck with backwards compatibility code (or an update method for bootstrap.yaml) when we change it later.

I'd say if it is not in a tagged release it never existed :laughing: Getting more serious, if there are no other things to change in this PR than the schema for `enable` I wonder if we can:
1. Rebase the #22720 branch on this one
2. Have another PR from that branch to this branch 
 
and still have CI running (since we are not targeting `develop`). If we can do that we can hold on on merging this to develop, and work in another PR whether we should go in one direction or the other since we have also bootstrapping from binaries available. What do you think?

Comment (becker33) [reply lib/spack/spack/schema/bootstrap.py:11]: I think it's fine as long as we don't have to write backwards compatibility code for it just because it's in develop for a little while.

**Review by becker33 — APPROVED** [review]
